jps/jps-builders/src/org/jetbrains/jps/incremental/instrumentation/NotNullInstrumentingBuilder.java

```diff
@@ -52,8 +50,10 @@ protected boolean isEnabled(CompileContext context, ModuleChunk chunk) {
   }
 
   @Override
-  protected boolean canInstrument(CompiledClass compiledClass, int classFileVersion) {
-    return (classFileVersion & 0xFFFF) >= Opcodes.V1_5 && !"module-info".equals(compiledClass.getClassName());
+  protected boolean canInstrument(CompiledClass compiledClass, ClassReader reader, int classFileVersion) {
+    return (classFileVersion & 0xFFFF) >= Opcodes.V1_5
+           && !"module-info".equals(compiledClass.getClassName())
+           && !IsKotlinClassCheckingVisitor.isKotlinClass(reader);
```

Comment:
This code will parse each class file to check presence of Kotlin annotation, and this looks inefficient. Also since this code is included in Java plugin these checks will be performed even if Kotlin plugin is disabled. And anyway it looks untidy when Java plugin includes Kotlin-specific code.

I think it would be better to do something in Kotlin plugin code which will turn off not-null instrumentation for a particular class file. E.g. you may put a flag to CompiledClass instance (it implements UserDataHolder) and check that flag in `NotNullInstrumentingBuilder`. This will be much more efficient and you'll get rid of Kotlin-specific code in Java plugin. What do you think?

Reply:
@altsvetkov ?